firefox-ios/Client/Utils/ContinuationsChecker.swift

```diff
@@ -25,3 +25,54 @@ struct ContinuationsChecker {
         return true
     }
 }
+
+final class SafeContinuation<T>: @unchecked Sendable {
+    private var continuation: UnsafeContinuation<T, Never>?
+    private var checkedContinuation: CheckedContinuation<T, Never>?
+    private var hasResumed = false
+
+    private let lock = NSLock()
+
+    private init(continuation: UnsafeContinuation<T, Never>) {
+        self.continuation = continuation
+    }
+
+    private init(checkedContinuation: CheckedContinuation<T, Never>) {
+        self.checkedContinuation = checkedContinuation
+    }
+
+    func resume(returning value: T) {
+        lock.lock()
+        defer { lock.unlock() }
+
+        guard !hasResumed else {
+            assertionFailure("⚠️ Continuation already resumed.")
+            return
+        }
+
+        hasResumed = true
+        if let c = continuation {
+            c.resume(returning: value)
+            continuation = nil
+        } else if let c = checkedContinuation {
+            c.resume(returning: value)
+            checkedContinuation = nil
+        } else {
+            assertionFailure("⚠️ No continuation available.")
+        }
+    }
+
+    static func run(unsafe: Bool = false, operation: @Sendable @escaping (SafeContinuation<T>) -> Void) async -> T {
+        if unsafe {
+            return await withUnsafeContinuation { c in
+                let wrapper = SafeContinuation(continuation: c)
+                operation(wrapper)
+            }
+        } else {
+            return await withCheckedContinuation { c in
+                let wrapper = SafeContinuation(checkedContinuation: c)
```

Comment:
nit; I think having one letter variables makes it a tiny bit harder to read
```suggestion
            return await withUnsafeContinuation { continuation in
                let wrapper = SafeContinuation(continuation: continuation)
                operation(wrapper)
            }
        } else {
            return await withCheckedContinuation { continuation in
               let wrapper = SafeContinuation(continuation: continuation)
```